### PR TITLE
Add more informative error messages when cypress verify hangs

### PR DESCRIPTION
**Pull request by stuartsan**
address #819

The purpose of this change is to add better visibility when `cypress run` hangs indefinitely on the verification step.

When running the smoke test, the child process's stderr contains useful information but it is currently being swallowed when the process doesn't exit. The proposed change writes the child's stderr to the parent's.

The result is that instead of seeing this:

```
Verifying Cypress can run
```
With no further output, I now see this:
```
Verifying Cypress can run
...
A JavaScript error occurred in the main process
Uncaught Exception:
Error: Failed to get 'appData' path
    at Object.<anonymous> (/tmp/lib/resources/electron.asar/browser/init.js:149:39)
    at Object.<anonymous> (/tmp/lib/resources/electron.asar/browser/init.js:173:3)
    at Module._compile (module.js:569:30)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:503:32)
    at tryModuleLoad (module.js:466:12)
    at Function.Module._load (module.js:458:3)
    at Function.Module.runMain (module.js:605:10)
    at startup (bootstrap_node.js:167:16)
    at bootstrap_node.js:589:3
```
And then, after googling and fixing my specific underlying problem (Electron's default `appData` path, `~/.config`, was read-only), if I do `cypress run` again it works fine.

This doesn't address the underlying issue of why `cypress --smoke-test` seems to be not exiting when this kind of failure happens, but it at least adds visibility so folks can self-serve and fix their specific problem more easily.

See also: https://github.com/cypress-io/cypress/issues/819#issuecomment-475308665.

### Validation

To validate this, I've made the change manually to the cypress CLI source within my own project. I cloned the cypress repo and fiddled around but couldn't figure out a good way to test it within here. 

This doesn't seem like it lends itself well to an automated test (cause it would be hanging on purpose)...what I wanted to do manually to reproduce the problem case is something like:

```
mkdir readonly
chmod 444 readonly
```
And then (on linux) set `XDG_CONFIG_HOME` (per [this](https://electronjs.org/docs/api/app#appgetpathname)) to tell Electron to use that as the appData path, and do `cypress run` (or maybe `cypress verify` for a smaller case).

Anyway, I read the contributing docs but couldn't figure out how to do that within this repo. I'm hoping that between feedback from the CI tests, and maintainers, I can get an idea if this change is legit and if there's anything else I should do to validate.

### CODE REVIEW

File: cli/lib/tasks/verify.js [hunk 1]
```diff
@@ -65,7 +65,7 @@ const runSmokeTest = (binaryDir) => {
 
     debug('smoke test command:', smokeTestCommand)
 
-    return Promise.resolve(util.exec(cypressExecPath, args))
+    return Promise.resolve(util.exec(cypressExecPath, args, { stdio: ['pipe', 'pipe', process.stderr] }))
```
Comment (stuartsan) [inline cli/lib/tasks/verify.js]: One specific concern I have but am unsure how to test is, the `onSmokeTestError` callback contains:
```js
return throwFormErrorText(errors.missingDependency)(err.stderr || err.message)
```
And possibly the suggested change is hosing its ability to hook into `err.stderr`, for those cases where the smoke test exist with useful info on `stderr`.

Comment (stuartsan) [reply cli/lib/tasks/verify.js]: Just pushed an update that addresses this

Comment (stuartsan) [reply cli/lib/tasks/verify.js]: Welp. The previous approach passed CI, but the latest doesn't.

Anybody have thoughts on whether the previous approach https://github.com/cypress-io/cypress/pull/3807/commits/fb331f231ff88e5a4371d31d26a06008c5b252cd seems ok?

File: cli/lib/tasks/verify.js [hunk 2]
```diff
@@ -65,7 +65,17 @@ const runSmokeTest = (binaryDir) => {
 
     debug('smoke test command:', smokeTestCommand)
 
-    return Promise.resolve(util.exec(cypressExecPath, args))
+    function smokeTestExec () {
+      let child = util.exec(cypressExecPath, args)
+
+      if (child.stderr) {
```
Comment (stuartsan) [inline cli/lib/tasks/verify.js]: Looking at https://github.com/sindresorhus/execa/tree/v0.10.0, no idea why sometimes `stderr` isn't there, but the unit tests fail without this guard 🤷‍♂️ .

**Comment by stuartsan** [comment]
Status: this is ready for review.

There is a failing integration test,

```
Warning: It looks like you are trying to record this run from a forked PR.
The 'Record Key' is missing. Your CI provider is likely not passing private environment variables to builds from forks.
```
But it doesn't seem related to the code change.

**Comment by jennifer-shehane** [comment]
Hey @stuartsan, we haven't forgotten about you. We will make sure to review this before our next release. 😄

**Comment by stuartsan** [comment]
Thanks @jennifer-shehane, all good, literally zero rush on my end, :)

**Comment by sebastianheise** [comment]
While not actually testing this PR, it did give the crucial tip to check for permissions on `~/.config/`. So this solved my `cypress verify` issue. More debug output is always good. 👍

**Comment by Bkucera** [comment]
I think adding a timeout to the smoke test process is a better solution here, we can then print you a proper error message and not hang the process forever. 

If the process does not exit within 10s, we'll kill it and print its `stderr

PR for adding the timeout here #4080

**Pull request closed without merging**